regions: updated with  new geometry and new centers | deleted old cen…

andyanderso opened the pull request:
**_Summary Commit Message:_**
regions: updated with new geometry and new centers | deleted old centers | updated center links || forecasts: added avy.org parsers for new and updated centers | updated data links || tests: added avy.org test files and code for updated centers.

**_Details:_**

**regions.js:**
_updated the following polygons with new geometry:_ 

- Northwest Avalanche Center - NWAC - Cascades West - Baker, Cascades East North, Cascaded West Central, Cascades East South, Snoqualmie Pass
- Wallowa - WAC - All zones
- Flathead - FAC - Flathead and Glacier NP zone
- Idaho Panhandle Avalanche Center - IPAC - St. Regis / Silver Valley and Kootenai (switched from FAC)
- Utah Avalanche Center - UAC - All zones
- Mt. Washington Avalanche Center - MWAC - Northern Presidential Range
- Valdez - VAC- All zones
- Kachina Peaks - KPAC- All zones
- Sawtooth - SNFAC - All Zones
- Chugach NF Avalanche Information Center - CNFAIC - All Zones
- Cordova - All zones
- Juneau - All Zones

_Added the following centers with polygons:_

- Central Oregon Avalanche Center - COAA - All zones
- Gallatin NF Avalanche Center - GNFAC - Added Centennial Range
- Taos- TAC - Added center and zone
- Utah Avalanche Center - UAC - Added Southwest Region
- Bridgeport Avalanche Center - BAC - Added zone
- Eastern AK Range - AAIC - Added zone
- CNFAIC - added Seward and Lost Lake Zone and Chugach State park

_Deleted the following centers:_

- Anchorage Avalanche Center - they shut down this year
- Crested Butte Avalanche Center - for consistency with the national map

_Updated the links where they had changed and to https where appropriate_
**forecasts.js:**

- updated to https links where appropriate
- Switched to using avy.org data files and avy.org parser for the following centers: SNFAC, UAC, AAIC, CNFAIC, COAA, BAC, GNFAC, TAC, MWAC
- Cleaned up the code by deleting old regions
- Switched Kootenai from FAC to IPAC

**Test folder:**

- Added test files and folders for the updated centers using avy.org json files
- Added tests to target centers using avy.org

spollack commented on forecasts.js:
```diff
@@ -366,75 +348,84 @@ forecasts.getRegionDetailsForRegionId = function(regionId) {
+                        2: 'https://www.idahopanhandleavalanche.org/advisories/selkirks-cabinets-map-xml',
+                        3: 'https://www.idahopanhandleavalanche.org/advisories/st-regis-silver-map-xml',
+                        4: 'https://www.idahopanhandleavalanche.org/advisories/kootenai-map-xml',
+                        5: 'https://www.idahopanhandleavalanche.org/advisories/kootenai-map-xml'
```
did kootenai move from FAC to IPAC?

andyanderso replied:
Yes. They moved over in the fall of 2017.

spollack replied:
perfect, thanks.

spollack commented on forecasts.js:
```diff
@@ -235,21 +234,12 @@ forecasts.forecastForRegionId = function(regionId, onForecast) {
-        process.nextTick(function() { onForecast(forecast); } );
-    } else {
+    }
+    else {
```
code style nit: let's have the two lines above on one line, i.e. 
`} else {`
to match the style of the rest of the code.

andyanderso replied:
Sounds great. Sorry about that.

spollack commented on forecasts.js:
```diff
@@ -271,18 +261,18 @@ forecasts.forecastForRegionId = function(regionId, onForecast) {
                         forecasts.mostRecentForecasts[regionId] = forecast;
                     }
-                    
+
```
thanks for cleaning up all this extra whitespace

andyanderso replied:
no worries.

spollack commented on forecasts.js:
```diff
@@ -306,58 +296,50 @@ forecasts.getRegionDetailsForRegionId = function(regionId) {
                     break;
                 case 'cacb': // CAC blog-only forecasts, which we don't parse
-                    dataURL = 'http://www.avalanche.ca/blogs?category=' + components[1];
+                    dataURL = 'https://www.avalanche.ca/blogs?category=' + components[1];
```
thanks for moving all these to https. 

have you verified that all these URLs resolve correctly?  thanks

andyanderso replied:
I verified all of them. The only center not using https right now is COAA
Of course, it would not hurt to have another person check them too :)

spollack replied:
great!

spollack commented:
great work Andy! i just took a first read, and put in a few minor comments. 

next i will check out your branch locally and take it for a test spin. i'll post back here with anything further i find.

spollack commented:
i checked out your repo locally, and there are failing tests. do you see these same failures? 

```
 65 passing (240ms)
  6 failing

  1) parseForecast_jac file001.html should return the correct forecast details:

      AssertionError [ERR_ASSERTION]: expected 0 to be 2
      + expected - actual

      +2
      -0
      
      at new AssertionError (internal/assert.js:268:11)
      at Assertion.assert (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:134:15)
      at Assertion.prop.(anonymous function) (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:61:14)
      at context.<anonymous> (/Users/spollack/temp/AviForecastWS/test/test.js:756:41)
      at callFn (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:249:21)
      at Test.Runnable.run (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:242:7)
      at Runner.runTest (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:373:10)
      at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:451:12
      at next (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:298:14)
      at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:308:7

  2) parseForecast_jac file002.html should return the correct forecast details:

      AssertionError [ERR_ASSERTION]: expected 0 to be 1
      + expected - actual

      +1
      -0
      
      at new AssertionError (internal/assert.js:268:11)
      at Assertion.assert (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:134:15)
      at Assertion.prop.(anonymous function) (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:61:14)
      at context.<anonymous> (/Users/spollack/temp/AviForecastWS/test/test.js:767:41)
      at callFn (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:249:21)
      at Test.Runnable.run (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:242:7)
      at Runner.runTest (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:373:10)
      at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:451:12
      at next (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:298:14)
      at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:308:7

  3) parseForecast_jac file003.html should return the correct forecast details:

      AssertionError [ERR_ASSERTION]: expected 0 to be 1
      + expected - actual

      +1
      -0
      
      at new AssertionError (internal/assert.js:268:11)
      at Assertion.assert (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:134:15)
      at Assertion.prop.(anonymous function) (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:61:14)
      at context.<anonymous> (/Users/spollack/temp/AviForecastWS/test/test.js:778:41)
      at callFn (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:249:21)
      at Test.Runnable.run (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:242:7)
      at Runner.runTest (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:373:10)
      at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:451:12
      at next (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:298:14)
      at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:308:7

  4) parseForecast_jac file004.html should return the correct forecast details:

      AssertionError [ERR_ASSERTION]: expected 0 to be 2
      + expected - actual

      +2
      -0
      
      at new AssertionError (internal/assert.js:268:11)
      at Assertion.assert (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:134:15)
      at Assertion.prop.(anonymous function) (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:61:14)
      at context.<anonymous> (/Users/spollack/temp/AviForecastWS/test/test.js:789:41)
      at callFn (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:249:21)
      at Test.Runnable.run (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:242:7)
      at Runner.runTest (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:373:10)
      at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:451:12
      at next (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:298:14)
      at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:308:7

  5) parseForecast_jac file005.html should return the correct forecast details:

      AssertionError [ERR_ASSERTION]: expected 0 to be 1
      + expected - actual

      +1
      -0
      
      at new AssertionError (internal/assert.js:268:11)
      at Assertion.assert (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:134:15)
      at Assertion.prop.(anonymous function) (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:61:14)
      at context.<anonymous> (/Users/spollack/temp/AviForecastWS/test/test.js:800:41)
      at callFn (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:249:21)
      at Test.Runnable.run (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:242:7)
      at Runner.runTest (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:373:10)
      at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:451:12
      at next (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:298:14)
      at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:308:7

  6) parseForecast_jac file006.html should return the correct forecast details:

      AssertionError [ERR_ASSERTION]: expected 0 to be 1
      + expected - actual

      +1
      -0
      
      at new AssertionError (internal/assert.js:268:11)
      at Assertion.assert (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:134:15)
      at Assertion.prop.(anonymous function) (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:61:14)
      at context.<anonymous> (/Users/spollack/temp/AviForecastWS/test/test.js:811:41)
      at callFn (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:249:21)
      at Test.Runnable.run (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:242:7)
      at Runner.runTest (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:373:10)
      at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:451:12
      at next (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:298:14)
      at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:308:7
```

andyanderso commented:
Yes. Those are for the Juneau Center which I did not modify. It looks like
they are in the process of changing their website (it forwards to a beta
site) as well so I was thinking that we could hold off on fixing the JAC
until they finish updating so we don't have to fix it twice.

https://beta.juneau.org/emergency/current-advisory

We could also update the test files with new html files and see how that
works for now.

On Wed, Jan 15, 2020 at 9:32 AM Seth Pollack <notifications@github.com>
wrote:

> i checked out your repo locally, and there are failing tests. do you see
> these same failures?
>
>  65 passing (240ms)
>   6 failing
>
>   1) parseForecast_jac file001.html should return the correct forecast details:
>
>       AssertionError [ERR_ASSERTION]: expected 0 to be 2
>       + expected - actual
>
>       +2
>       -0
>
>       at new AssertionError (internal/assert.js:268:11)
>       at Assertion.assert (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:134:15)
>       at Assertion.prop.(anonymous function) (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:61:14)
>       at context.<anonymous> (/Users/spollack/temp/AviForecastWS/test/test.js:756:41)
>       at callFn (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:249:21)
>       at Test.Runnable.run (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:242:7)
>       at Runner.runTest (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:373:10)
>       at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:451:12
>       at next (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:298:14)
>       at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:308:7
>
>   2) parseForecast_jac file002.html should return the correct forecast details:
>
>       AssertionError [ERR_ASSERTION]: expected 0 to be 1
>       + expected - actual
>
>       +1
>       -0
>
>       at new AssertionError (internal/assert.js:268:11)
>       at Assertion.assert (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:134:15)
>       at Assertion.prop.(anonymous function) (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:61:14)
>       at context.<anonymous> (/Users/spollack/temp/AviForecastWS/test/test.js:767:41)
>       at callFn (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:249:21)
>       at Test.Runnable.run (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:242:7)
>       at Runner.runTest (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:373:10)
>       at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:451:12
>       at next (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:298:14)
>       at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:308:7
>
>   3) parseForecast_jac file003.html should return the correct forecast details:
>
>       AssertionError [ERR_ASSERTION]: expected 0 to be 1
>       + expected - actual
>
>       +1
>       -0
>
>       at new AssertionError (internal/assert.js:268:11)
>       at Assertion.assert (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:134:15)
>       at Assertion.prop.(anonymous function) (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:61:14)
>       at context.<anonymous> (/Users/spollack/temp/AviForecastWS/test/test.js:778:41)
>       at callFn (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:249:21)
>       at Test.Runnable.run (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:242:7)
>       at Runner.runTest (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:373:10)
>       at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:451:12
>       at next (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:298:14)
>       at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:308:7
>
>   4) parseForecast_jac file004.html should return the correct forecast details:
>
>       AssertionError [ERR_ASSERTION]: expected 0 to be 2
>       + expected - actual
>
>       +2
>       -0
>
>       at new AssertionError (internal/assert.js:268:11)
>       at Assertion.assert (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:134:15)
>       at Assertion.prop.(anonymous function) (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:61:14)
>       at context.<anonymous> (/Users/spollack/temp/AviForecastWS/test/test.js:789:41)
>       at callFn (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:249:21)
>       at Test.Runnable.run (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:242:7)
>       at Runner.runTest (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:373:10)
>       at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:451:12
>       at next (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:298:14)
>       at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:308:7
>
>   5) parseForecast_jac file005.html should return the correct forecast details:
>
>       AssertionError [ERR_ASSERTION]: expected 0 to be 1
>       + expected - actual
>
>       +1
>       -0
>
>       at new AssertionError (internal/assert.js:268:11)
>       at Assertion.assert (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:134:15)
>       at Assertion.prop.(anonymous function) (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:61:14)
>       at context.<anonymous> (/Users/spollack/temp/AviForecastWS/test/test.js:800:41)
>       at callFn (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:249:21)
>       at Test.Runnable.run (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:242:7)
>       at Runner.runTest (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:373:10)
>       at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:451:12
>       at next (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:298:14)
>       at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:308:7
>
>   6) parseForecast_jac file006.html should return the correct forecast details:
>
>       AssertionError [ERR_ASSERTION]: expected 0 to be 1
>       + expected - actual
>
>       +1
>       -0
>
>       at new AssertionError (internal/assert.js:268:11)
>       at Assertion.assert (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:134:15)
>       at Assertion.prop.(anonymous function) (/Users/spollack/temp/AviForecastWS/node_modules/should/lib/should.js:61:14)
>       at context.<anonymous> (/Users/spollack/temp/AviForecastWS/test/test.js:811:41)
>       at callFn (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:249:21)
>       at Test.Runnable.run (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runnable.js:242:7)
>       at Runner.runTest (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:373:10)
>       at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:451:12
>       at next (/Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:298:14)
>       at /Users/spollack/temp/AviForecastWS/node_modules/mocha/lib/runner.js:308:7
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/spollack/AviForecastWS/pull/9?email_source=notifications&email_token=ACEFMEMM2VIEPDEKJ5LZRODQ55CELA5CNFSM4KHGYGX2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEJBEOGI#issuecomment-574768921>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACEFMEKHSFED5TIA7QIESELQ55CELANCNFSM4KHGYGXQ>
> .
>


-- 
Andy Anderson
Instagram: @andyanders0 <https://www.instagram.com/andyanders0/>
Twitter: @andyanderso <https://twitter.com/andyanderso>
andyanderso@gmail.com

spollack commented:
@andyanderso those tests pass in master. since they run against saved test artifacts, they are not affected by changes to the live site. something in your change set broke these it appears. let me know if you need help tracking that down. but we need to fix that. thanks!

andyanderso commented:
copy that. I will see if I can track it down. I am off the computer till tonight so if you happen to find it before then let me know :)

spollack commented:
next issue: if i run the server `node main.js` and let it load up all the forecasts, on master right now there are 7 invalid forecasts, on your branch there are 26. the log line looks like:
`there were invalid forecasts; invalid forecast count: 26`
we need to run these down, and see if things went backwards for certain regions, and if so, fix them.

spollack commented:
i just verified that the new regions.json is a valid json file. but hard to validate that the regions are all set up correctly (correct geometry, names, links, etc.) without loading it into the app unfortunately. 

what i've done in the past is to generally make these type of changes before the season starts, so i can test in production without much risk. (not awesome but it works.) we don't have that luxury right now. 

i've also in the past set up test builds of the app that point to a staging or local test server to be able to do this type of verification. but that requires a special build of the app, or adding a special mode. so we would need GR to help with that. 

If we can't do that, we probably will need to just deploy to production (once all the other issues are fixed), quickly verify everything, and if there are any issues, roll back. sigh. 

@andyanderso  any other thoughts on how to verify?

andyanderso commented:
> @andyanderso those tests pass in master. since they run against saved test artifacts, they are not affected by changes to the live site. something in your change set broke these it appears. let me know if you need help tracking that down. but we need to fix that. thanks!

I found this one. it was an https replacing an http in line 1170 of forecasts.js.

andyanderso commented:
> node main.js

I will see about tracking those down tonight.

andyanderso commented:
> i just verified that the new regions.json is a valid json file. but hard to validate that the regions are all set up correctly (correct geometry, names, links, etc.) without loading it into the app unfortunately.
> 
> what i've done in the past is to generally make these type of changes before the season starts, so i can test in production without much risk. (not awesome but it works.) we don't have that luxury right now.
> 
> i've also in the past set up test builds of the app that point to a staging or local test server to be able to do this type of verification. but that requires a special build of the app, or adding a special mode. so we would need GR to help with that.
> 
> If we can't do that, we probably will need to just deploy to production (once all the other issues are fixed), quickly verify everything, and if there are any issues, roll back. sigh.
> 
> @andyanderso any other thoughts on how to verify?

Could we use your conversion tools to convert back to a geojson format then load them into a simple geojson viewer like this http://geojson.io?

andyanderso commented:
> next issue: if i run the server `node main.js` and let it load up all the forecasts, on master right now there are 7 invalid forecasts, on your branch there are 26. the log line looks like:
> `there were invalid forecasts; invalid forecast count: 26`
> we need to run these down, and see if things went backwards for certain regions, and if so, fix them.

@spollack I will see about tracking those down tonight.

spollack commented:
yeah i don't have a region reverse conversion tool written unfortunately.

andyanderso commented:
Ok - I had some more time to work on this stuff. Most of the invalid forecasts are triggering a warning due to those regions not issuing everyday or not issuing danger ratings or both. 

So we get UNEXPECTED got aviLevel 0 "date":"Invalid date"

I tried putting the regions into the "certain regions do not issue new forecasts daily, so this case can happen" section @ line 232 and into the "these regions currently do not provide any danger level ratings" @ line 126. 

They still trigger a warning. We could use the `parser = forecasts.parseForecast_noop;` parse for them but I am not sure that they won't ever issue danger ratings. So I think it is safer to leave the warnings and have them setup to parse danger ratings just in case those centers issue ratings. 

There were a few centers that were not pulling data - JAC, VIAC, HG. I updated the parser for the JAC and it works now. I have not had a chance to update the VIAC and HG ones. They look like they are using something similar. Do you have time to take a look at those two?

Then the other two errors were just typos on my part (Taos and HPAC).

So now I am pretty sure we have everything pulling the data it is supposed to.

spollack commented on forecasts.js:
```diff
@@ -18,6 +18,7 @@ var cheerio = require('cheerio');
 
 
 // avi danger levels
+forecasts.AVI_LEVEL_UNKNOWN = -1;
```
this doesn't look right -- we overwrite this value with zero in the next line. i think we want to leave this as zero. i.e. we don't need this new line of code.

andyanderso replied:
I put it in there because avy.org is showing -1 as a danger rating. I am not sure if they will ever use 0 in place of -1 nor am I sure if there would ever be a difference. Regardless I think you are right for now it really is not needed and if someday there are two different ratings of -1 and 0 we will figure that out then. Thanks for making me think more about it.

andyanderso commented:
I had more time today so I started updating the VIAC and the HG parsers. I not particularly good at regex but I got dates and days of the week working. I think I have good expressions to get the danger ratings but they are not parsing right. So those need more work. I am out of time for today but I figured I would share my work in case you have time to clean it up.

spollack commented:
Hi Andy. just getting a moment to look at this. thanks for your work here. 

i see in the current state, we have failing unit tests for VIAC, HG, and JAC. i see you are working on VIAC and HG, so that is still in flight. for JAC, you said you had updated the parser right, so you may just need to update the tests to modern test data? 

running it, at this moment we have 43 invalid forecasts out of the 116 total. i didn't dig into the details yet.

so yeah a bit more work here to drive both of these counts to zero.

spollack commented:
by the way, for the future, it's often better to do the work in more bite-sized PRs that we can merge and ship as each is ready, vs. one large one, as any one issue in a big PR keeps the whole set of improvements from being clear to ship. no worries for this one though. (unless you feel like its worth splitting into a couple of pieces, that would be fine too.)

spollack commented:
it looks like we are getting in invalid date for a bunch of regions. that is the cause of many of the 43 invalid forecasts. 

for example: `UNEXPECTED got aviLevel 0 in forecast; regionId: coaa_205_1; forecast: [{"date":"Invalid date","aviLevel":0}]`

spollack commented:
ahh, the issue here is that a bunch of these forecasts are coming from the avalanche.org API with a null value for the forecast date. 

For example, the response from `https://api.avalanche.org/v1/forecast/get-map-data/COAA` right now is: 

`{"type":"FeatureCollection","features":[{"type":"Feature","id":205,"properties":{"name":"Central Oregon Mountains","center":"Central Oregon Avalanche Association","timezone":"America\/Los_Angeles","state":"OR","rating":null,"travel_advice":"Watch for signs of unstable snow such as recent avalanches, cracking in the snow, and audible collapsing. Avoid traveling on or under similar slopes.","danger":"no rating","danger_level":-1,"color":"#888888","stroke":"#104efb","font_color":"#ffffff","link":"http:\/\/www.coavalanche.org\/","start_date":null,"end_date":null,"warning":{"product":null,"start_date":null,"end_date":null,"reason":null}},"geometry":{"type":"MultiPolygon","coordinates":[[[[-121.23025,43.76576],[-121.23514,43.76652],[-121.25347,43.76785],[-121.28012,43.76724],[-121.29317,43.76344],[-121.29664,43.76161],[-121.3001,43.75801],[-121.30346,43.7509],[-121.30574,43.74733],[-121.30552,43.74206],[-121.30655,43.73764],[-121.30877,43.73319],[-121.31331,43.72604],[-121.31539,43.71808],[-121.3146,43.70051],[-121.3117,43.69002],[-121.31,43.6795],[-121.3074,43.67603],[-121.30212,43.66648],[-121.29828,43.66218],[-121.28457,43.65458],[-121.27719,43.65214],[-121.26977,43.64882],[-121.2661,43.64803],[-121.26115,43.64553],[-121.2343,43.64009],[-121.2295,43.64024],[-121.19297,43.63583],[-121.16978,43.63624],[-121.15032,43.63922],[-121.137,43.6421],[-121.11412,43.65046],[-121.10946,43.65583],[-121.10373,43.66475],[-121.09702,43.68071],[-121.09235,43.68696],[-121.09019,43.69404],[-121.09072,43.70723],[-121.09221,43.71336],[-121.09489,43.71859],[-121.09619,43.72033],[-121.09767,43.72646],[-121.10025,43.7308],[-121.1116,43.73936],[-121.12424,43.74526],[-121.1454,43.75448],[-121.14909,43.75527],[-121.15403,43.7578],[-121.16767,43.7619],[-121.17512,43.76349],[-121.19374,43.76395],[-121.20473,43.76721],[-121.20479,43.76897],[-121.23025,43.76576]]],[[[-121.82129,44.72046],[-121.82782,44.71902],[-121.8329,44.71485],[-121.84062,44.70518],[-121.84571,44.70182],[-121.84953,44.69847],[-121.85081,44.6968],[-121.86203,44.6906],[-121.86758,44.68558],[-121.87077,44.68161],[-121.87576,44.66882],[-121.87695,44.6645],[-121.8776,44.6555],[-121.8772,44.65349],[-121.88029,44.64441],[-121.87994,44.63649],[-121.88163,44.6299],[-121.882,44.62332],[-121.88336,44.61995],[-121.88492,44.61756],[-121.88475,44.61318],[-121.88601,44.6021],[-121.88696,44.59762],[-121.88704,44.59338],[-121.88967,44.58824],[-121.88952,44.57783],[-121.89162,44.57152],[-121.89287,44.5631],[-121.89188,44.55231],[-121.88916,44.55056],[-121.88532,44.54975],[-121.88549,44.5372],[-121.88497,44.53097],[-121.88609,44.52916],[-121.88493,44.51494],[-121.88336,44.50882],[-121.88055,44.50359],[-121.88009,44.49479],[-121.87719,44.48694],[-121.87516,44.44908],[-121.87511,44.44118],[-121.87758,44.43319],[-121.87724,44.42612],[-121.87815,44.42167],[-121.87786,44.41814],[-121.87847,44.41017],[-121.87809,44.40577],[-121.88029,44.40218],[-121.88134,44.3995],[-121.88101,44.39245],[-121.88214,44.38976],[-121.88314,44.38443],[-121.88292,44.38091],[-121.88599,44.36756],[-121.88581,44.36224],[-121.88692,44.35956],[-121.88783,44.35424],[-121.88763,44.3516],[-121.88843,44.34541],[-121.8881,44.34103],[-121.88894,44.33573],[-121.8881,44.32255],[-121.88668,44.31996],[-121.88506,44.31384],[-121.88456,44.30504],[-121.88556,44.3006],[-121.88518,44.29445],[-121.88613,44.28822],[-121.88531,44.27416],[-121.88125,44.26902],[-121.87842,44.26384],[-121.87578,44.26129],[-121.87302,44.257],[-121.86916,44.25448],[-121.86166,44.25295],[-121.85286,44.25058],[-121.84665,44.24989],[-121.80969,44.251],[-121.80113,44.25305],[-121.78383,44.25363],[-121.77536,44.25652],[-121.76155,44.25719],[-121.75537,44.25925],[-121.7504,44.26038],[-121.7444,44.26411],[-121.7378,44.27187],[-121.73566,44.27533],[-121.73583,44.27792],[-121.73499,44.28312],[-121.73284,44.28659],[-121.73195,44.29092],[-121.72993,44.29613],[-121.7266,44.30051],[-121.72351,44.30839],[-121.71898,44.31555],[-121.71231,44.32896],[-121.70917,44.33873],[-121.70969,44.34752],[-121.71242,44.35187],[-121.72027,44.35961],[-121.72299,44.36396],[-121.72443,44.36746],[-121.72967,44.37262],[-121.7311,44.37612],[-121.73372,44.37871],[-121.73516,44.38221],[-121.73929,44.39007],[-121.74313,44.39441],[-121.74573,44.39967],[-121.74707,44.4014],[-121.74986,44.40663],[-121.75407,44.41181],[-121.75811,44.4179],[-121.76087,44.42579],[-121.76098,44.42933],[-121.76253,44.43547],[-121.76303,44.44343],[-121.76085,44.44793],[-121.75985,44.4515],[-121.75764,44.456],[-121.7566,44.45957],[-121.75685,44.46399],[-121.75583,44.46757],[-121.75497,44.47294],[-121.75532,44.48089],[-121.75426,44.48446],[-121.75514,44.50299],[-121.7568,44.51263],[-121.75857,44.51804],[-121.75998,44.52411],[-121.76179,44.54152],[-121.76332,44.54933],[-121.76099,44.5529],[-121.75438,44.56641],[-121.75012,44.57557],[-121.74911,44.57918],[-121.748,44.58101],[-121.74495,44.59684],[-121.74391,44.60458],[-121.74186,44.6101],[-121.74216,44.61294],[-121.74171,44.62963],[-121.74055,44.63223],[-121.74001,44.63978],[-121.73895,44.64252],[-121.73802,44.64954],[-121.73582,44.65411],[-121.73602,44.65766],[-121.73376,44.66214],[-121.73268,44.6665],[-121.73364,44.68662],[-121.73689,44.69752],[-121.73736,44.70237],[-121.74169,44.71215],[-121.74788,44.71592],[-121.7639,44.72167],[-121.77621,44.7244],[-121.78375,44.72597],[-121.78987,44.7257],[-121.79129,44.72673],[-121.82129,44.72046]]],[[[-121.78121,44.19591],[-121.75029,44.19638],[-121.73779,44.19231],[-121.72695,44.191],[-121.69867,44.18576],[-121.69715,44.18135],[-121.68114,44.1749],[-121.67371,44.17251],[-121.66076,44.16475],[-121.65679,44.16126],[-121.64905,44.15611],[-121.64533,44.15447],[-121.63631,44.14847],[-121.62874,44.14512],[-121.6261,44.14252],[-121.61966,44.13737],[-121.6081,44.12966],[-121.60172,44.12626],[-121.59426,44.12297],[-121.59114,44.11241],[-121.59058,44.1027],[-121.5888,44.09304],[-121.58727,44.08779],[-121.58817,44.08247],[-121.58796,44.07895],[-121.58893,44.07366],[-121.58881,44.07103],[-121.5911,44.06746],[-121.5946,44.06386],[-121.60298,44.06013],[-121.61354,44.05107],[-121.61924,44.04213],[-121.6226,44.03501],[-121.62489,44.03232],[-121.62466,44.02881],[-121.62664,44.02173],[-121.62625,44.0156],[-121.62308,44.00252],[-121.62166,43.99817],[-121.61897,43.99297],[-121.61509,43.9869],[-121.61243,43.98169],[-121.61097,43.97734],[-121.61079,43.97294],[-121.61176,43.96495],[-121.61143,43.95611],[-121.61358,43.94809],[-121.61557,43.93743],[-121.61532,43.93124],[-121.61617,43.92327],[-121.61722,43.91971],[-121.61812,43.91262],[-121.61917,43.90906],[-121.62596,43.89649],[-121.62931,43.88932],[-121.63048,43.88841],[-121.6327,43.88394],[-121.63975,43.87848],[-121.65043,43.87386],[-121.6575,43.87021],[-121.66225,43.86836],[-121.67411,43.86464],[-121.68138,43.86174],[-121.69348,43.85784],[-121.70323,43.85576],[-121.70938,43.85557],[-121.71918,43.85437],[-121.74377,43.8536],[-121.75732,43.85408],[-121.76227,43.85482],[-121.77832,43.85611],[-121.78781,43.85599],[-121.79894,43.85743],[-121.8051,43.85723],[-121.81876,43.85947],[-121.82842,43.86195],[-121.8308,43.86192],[-121.84699,43.86674],[-121.85206,43.87012],[-121.85341,43.87274],[-121.85602,43.8762],[-121.86101,43.87782],[-121.86241,43.88221],[-121.86407,43.89278],[-121.86443,43.90072],[-121.86355,43.90781],[-121.86396,43.91662],[-121.86219,43.93168],[-121.85919,43.94587],[-121.85952,43.95025],[-121.85745,43.95468],[-121.85777,43.95906],[-121.8547,43.96616],[-121.85377,43.9697],[-121.85046,43.97419],[-121.84729,43.98132],[-121.8463,43.98487],[-121.84651,43.98839],[-121.84539,43.99196],[-121.84551,43.99638],[-121.83849,44.00717],[-121.83624,44.0134],[-121.83304,44.02759],[-121.83084,44.03469],[-121.82851,44.03915],[-121.829,44.04969],[-121.83198,44.05753],[-121.83225,44.06193],[-121.83627,44.06623],[-121.842,44.07404],[-121.84501,44.07752],[-121.84549,44.08195],[-121.84842,44.09069],[-121.84863,44.0951],[-121.85258,44.09853],[-121.87427,44.11292],[-121.87681,44.11636],[-121.87704,44.12512],[-121.87593,44.12779],[-121.87136,44.13409],[-121.86415,44.13955],[-121.856,44.14424],[-121.85147,44.14621],[-121.84694,44.14906],[-121.84089,44.1536],[-121.83616,44.16072],[-121.83385,44.16517],[-121.83019,44.16874],[-121.82659,44.17319],[-121.82304,44.17852],[-121.81592,44.18312],[-121.81239,44.18586],[-121.80909,44.19128],[-121.8028,44.19225],[-121.79771,44.1923],[-121.79143,44.19327],[-121.78121,44.19591]]]]}}]}`

@andyanderso this seems like a bug in the avalanche.org API. do you agree? if so, would you follow up with them please?

spollack commented:
or, if this is expected behavior, we need to understand how to interpret it, and modify our code appropriately.

andyanderso commented:
I sent Clark a quick note about the null values in the avy.org geojson files. The way I read it is that this is his way of accounting for the case of centers not issuing a danger rating or not issuing every day (or both). If they don't have a current product they get a null date value and a grey polygon.

I think our code is doing the same thing right now except that we get a warning when the center does not have a rating or a current date.

We could use the `parser = forecasts.parseForecast_noop;`  for these centers but I am not sure that they won't ever issue danger ratings. So I think it is safer to leave the warnings and have them set up to parse danger ratings just in case those centers issue ratings.

We could also create an exception in our code that says it is ok to get a null value for the date for certain centers but it seems like all of the centers that use avy.org for data will give us a null value eventually (summer) so we would have to have them all in that exception.

Once we figure out what to do with this that will take care of most of the warnings since we have so many centers on avy.org now. 

Other things:
Parsers:

- VIAC and HG parsers are works in progress. If you get a chance it would be great if you wanted to take a look at that code. I will try to do some more work on them this afternoon. 
- JAC should be good to go.

Tests:

- I will try to add new test files for VIAC, HG, and JAC this afternoon.

andyanderso commented:
I have not heard back from Clark yet so I went ahead and added the "grey" avy.org centers to our exceptions clauses. I changed the formate on those exception clauses to each region on a new line to make it easier to see how many centers are in the exceptions list.

We still get warnings for centers that have closed (like ipac) and any centers that have not posted for the day when you run the main.js. 

I also started working on the tests. So far still have 4 failing. I just wanted to push the new exception lists for now as a little push.

spollack commented:
@andyanderso i added an explicit check for null dates in parseForecast_avalanche_org_api

spollack commented:
@andyanderso i've temporarily disabled the parsers for JAC, VIAC, HG, for several reasons: 1) its hard to get these HTML parsers robust in general, 2) its harder out of season when forecasts aren't being issued, 3) its even harder when websites have unusual COVID-related changes.

spollack commented:
@andyanderso we'll need to fix up the "UNEXPECTED" forecast checking once the coming forecast season is underway. tricky right now.

spollack commented:
merged.
